django/db/models/base.py

```diff
@@ -909,6 +909,10 @@ def _save_parents(self, cls, using, update_fields):
         """Save all the parents of cls using values from self."""
         meta = cls._meta
         inserted = False
+
+        if self.pk is None and self._state.adding is True:
+            self._state.related_managers_cache = {}
```

Comment:
I'm not sure :thinking: this may be too simple. For example, everything works fine on Django 4.1 for inherited many-to-many fields:
```diff
diff --git a/tests/model_inheritance/models.py b/tests/model_inheritance/models.py
index a479faeda5..ba0ad5ed40 100644
--- a/tests/model_inheritance/models.py
+++ b/tests/model_inheritance/models.py
@@ -109,6 +109,8 @@ class ItalianRestaurant(Restaurant):
 class Supplier(Place):
     customers = models.ManyToManyField(Restaurant, related_name="provider")
 
+class NewSupplier(Supplier):
+    pass
 
 class ParkingLot(Place):
     # An explicit link to the parent (we can control the attribute name).
diff --git a/tests/model_inheritance/tests.py b/tests/model_inheritance/tests.py
index 882f2999e2..8a7c09800d 100644
--- a/tests/model_inheritance/tests.py
+++ b/tests/model_inheritance/tests.py
@@ -22,6 +22,7 @@ from .models import (
     Student,
     SubBase,
     Supplier,
+    NewSupplier,
     Title,
     Worker,
 )
@@ -148,6 +149,18 @@ class ModelInheritanceTests(TestCase):
         # accidentally found).
         self.assertQuerysetEqual(s.titles.all(), [])
 
+    def test_2(self):
+        r1 = Restaurant.objects.create()
+        ns = NewSupplier.objects.create(name="x", address="y")
+        ns.customers.set([r1])
+        old_customers = ns.customers.all()
+        ns.pk = None
+        ns.id = None
+        ns._state.adding = True
+        ns.save()
+        ns.customers.set(old_customers)
+        self.assertCountEqual(ns.customers.all(), old_customers)
+
     def test_update_parent_filtering(self):
         """
         Updating a field of a model subclass doesn't issue an UPDATE
```

and your solution adds unnecessary reset for `related_managers_cache`.

Reply:
resetting ```related_managers_cache``` is for issue related to cloning instances.

Reply:
The main difference is that `NewSupplier` inherits a `ManyToManyField`.

Reply:
I'm not sure if it's an acceptable solution. I'm afraid of introducing a performance regression :chart_with_upwards_trend:  as it basically creates an empty dictionary and `related_managers_cache` variable for all newly created objects, regardless of having or not cached managers.

Reply:
What if we replace ```self._state.related_managers_cache = {}``` with ```self._state.related_managers_cache.clear()```? It wont create an empty dictionary then.

Reply:
Agreed, let's try with `clear()`.

Reply:
I pushed extra tests to #16135.